False positive ``used-before-assignment``  for walrus operator different lines

After the support of the walrus operator, I've noticed that there's still a  `used-before-assignment` false positive when working on different lines:

### Steps to reproduce

Given a file `good-test.py`:
```python
"""Good test example."""
var = f'The number {(count := 4)} is equal to {count}'

```
`pylint` doesn't detect any errors which is expected. However when return to a line like this:

Given a file `bad-test.py`:
```python
"""Bad test example."""
var = f'The number {(count := 4)} ' \
      f'is equal to {count}'

```
We obtain the following:
### Current behavior

Result of `pylint good-test.py`:
```
************* Module bad-test
bad-test.py:2:7: E0601: Using variable 'count' before assignment (used-before-assignment)
```

### Expected behavior

Normally, this should not raise any errors.

### pylint --version output

Result of pylint --version output on my local macOS:

```
pylint 2.7.2
astroid 2.5.1
Python 3.8.8 (v3.8.8:024d8058b0, Feb 19 2021, 08:48:17) 
[Clang 6.0 (clang-600.0.57)]
```

--- a/ChangeLog
+++ b/ChangeLog
@@ -71,6 +71,10 @@ Release date: TBA
 
   Closes #4218
 
+* Improve handling of assignment expressions, better edge case handling
+
+  Closes #3763, #4238
+
 
 What's New in Pylint 2.7.2?
 ===========================

--- a/pylint/checkers/variables.py
+++ b/pylint/checkers/variables.py
@@ -59,6 +59,7 @@ from astroid.context import InferenceContext
 
 from pylint.checkers import BaseChecker, utils
 from pylint.checkers.utils import is_postponed_evaluation_enabled
+from pylint.constants import PY39_PLUS
 from pylint.interfaces import HIGH, INFERENCE, INFERENCE_FAILURE, IAstroidChecker
 from pylint.utils import get_global_option
 
@@ -1410,7 +1411,15 @@ class VariablesChecker(BaseChecker):
                     # same line as the function definition
                     maybee0601 = False
                 elif (
-                    isinstance(defstmt, astroid.Assign)
+                    isinstance(
+                        defstmt,
+                        (
+                            astroid.Assign,
+                            astroid.AnnAssign,
+                            astroid.AugAssign,
+                            astroid.Expr,
+                        ),
+                    )
                     and isinstance(defstmt.value, astroid.IfExp)
                     and frame is defframe
                     and defframe.parent_of(node)
@@ -1433,6 +1442,18 @@ class VariablesChecker(BaseChecker):
                             and defnode.col_offset < node.col_offset
                         )
                         or (defnode.lineno < node.lineno)
+                        or (
+                            # Issue in the `ast` module until py39
+                            # Nodes in a multiline string have the same lineno
+                            # Could be false-positive without check
+                            not PY39_PLUS
+                            and defnode.lineno == node.lineno
+                            and isinstance(
+                                defstmt,
+                                (astroid.Assign, astroid.AnnAssign, astroid.AugAssign),
+                            )
+                            and isinstance(defstmt.value, astroid.JoinedStr)
+                        )
                     )
                 ):
                     # Expressions, with assignment expressions

--- a/pylint/constants.py
+++ b/pylint/constants.py
@@ -8,6 +8,7 @@ from astroid.__pkginfo__ import version as astroid_version
 from pylint.__pkginfo__ import version as pylint_version
 
 PY38_PLUS = sys.version_info[:2] >= (3, 8)
+PY39_PLUS = sys.version_info[:2] >= (3, 9)
 PY310_PLUS = sys.version_info[:2] >= (3, 10)
 
 

--- a/tests/functional/a/assignment_expression.py
+++ b/tests/functional/a/assignment_expression.py
@@ -8,6 +8,9 @@ else:
     x = False
 
 x = b if (b := True) else False
+x2: bool = b2 if (b2 := True) else False
+x3 = 0
+x3 += b3 if (b3 := 4) else 6
 
 a = ["a   ", "b   ", "c   "]
 c = [text for el in a if (text := el.strip()) == "b"]
@@ -47,6 +50,24 @@ function = lambda: (
     i := h,
 )
 print(function())
+
+
+# https://github.com/PyCQA/pylint/issues/3763
+foo if (foo := 3 - 2) > 0 else 0  # [pointless-statement]
+
+
+# https://github.com/PyCQA/pylint/issues/4238
+l1 = f'The number {(count1 := 4)} ' \
+     f'is equal to {count1}'
+l2: str = (
+    f'The number {(count2 := 4)} '
+    f'is equal to {count2}'
+)
+l3 = "Hello "
+l3 += (
+    f'The number {(count3 := 4)} '
+    f'is equal to {count3}'
+)
 
 
 # check wrong usage

--- a/tests/functional/a/assignment_expression.txt
+++ b/tests/functional/a/assignment_expression.txt
@@ -1,3 +1,4 @@
-used-before-assignment:53:7::Using variable 'err_a' before assignment
-used-before-assignment:54:6::Using variable 'err_b' before assignment
-used-before-assignment:56:13::Using variable 'err_d' before assignment
+pointless-statement:56:0::Statement seems to have no effect
+used-before-assignment:74:7::Using variable 'err_a' before assignment
+used-before-assignment:75:6::Using variable 'err_b' before assignment
+used-before-assignment:77:13::Using variable 'err_d' before assignment

